### PR TITLE
dialects: add Snitch Runtime dialect

### PR DESCRIPTION
**Pull request by JosseVanDelm**
This is the snitch runtime dialect, that models calls present in 
https://github.com/pulp-platform/snitch/tree/b9fe5550e26ea878fb734cfc37d161f564252305/sw/snRuntime

Please note, this is working from commit `b9fe555` in the snitch repo, as currently there are some changes happening upstream!

Ops are here: https://github.com/pulp-platform/snitch/blob/b9fe5550e26ea878fb734cfc37d161f564252305/sw/snRuntime/include/snrt.h

### CODE REVIEW

**Review by AntonLydike — APPROVED** [review]

File: tests/dialects/test_snitch_runtime.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+from xdsl.dialects import snitch_runtime
+from xdsl.utils.test_value import TestSSAValue
+from xdsl.dialects.builtin import IntegerType, Signedness, IndexType
+
+u32 = IntegerType(data=32, signedness=Signedness.UNSIGNED)
+
+
+def test_snrt_op():
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply tests/dialects/test_snitch_runtime.py:8]: I'd go as far as to say that this is not a test, and should be a filecheck instead :) But my assumption is that this is an aspect of this PR still being WIP

**Comment by JosseVanDelm** [comment]
Thanks @webmiche! Will work on filecheck tests soon!
And yes, some smaller modular tests would be nice!
I'm just sketching as I go at the moment. I was thinking of making this into multiple smaller PR's after we have agreed on some basic stuff.

The plan is to run this stuff on `snitch_cluster`.
There are two options to that:
* Run the binary in a functional simulator called [banshee](https://github.com/pulp-platform/snitch/tree/master/sw/banshee) - fast, but we can not get real-world performance numbers
* Run the binary in a [simulation of the entire cluster - hardware simulator - verilator](https://github.com/pulp-platform/snitch/tree/master/hw/system/snitch_cluster) - slow, but we can get real-world performance numbers

I also have some colleagues working on smaller versions of the snitch-cluster, because now we would be starting with multi-core tests from the get-go, which doesn't make a lot of sense :smile:

**Comment by webmiche** [comment]
In what order of magnitude does fast live? I am thinking of incorporating this into the CI, mainly because it gives an easy way for me to understand whether it works and makes sense.
Though in the end, I probably won't review PRs on this, so not sure that matters really :sweat_smile:

**Comment by JosseVanDelm** [comment]
@webmiche Here are the full details:
https://pulp-platform.org/docs/Banshee_ICCAD_2021.pdf
@nazavode might have a more practical perspective on this

You can also see the CI over at the snitch repo :wink:

**Comment by webmiche** [comment]
My main concern is slowing down our work on the PRs because we are waiting for the CI. That being said, everything under 10 minutes is certainly okay, IMO. Currently, the MLIR flow takes a couple of minutes already, so it feels like there is room for a snitch CI.

**Comment by JosseVanDelm** [comment]
Removing the `WIP` as per @tobiasgrosser comment on Zulip

File: tests/dialects/test_snitch_runtime.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply tests/dialects/test_snitch_runtime.py]: I've removed this file and added a simple filecheck test

**Review by superlopuh — APPROVED** [review]

**Comment by JosseVanDelm** [comment]
@AntonLydike  are we good for merging?

**Review by AntonLydike — APPROVED** [review]

**Review by nazavode — APPROVED** [review]
Beautiful, thanks!

**Pull request merged**